### ISSUE TITLE
Rollup adds $0 to undefined

### ISSUE THREAD

**Issue by calebeby** [rollup/rollup#2290]
https://rollupjs.org/repl?version=0.61.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmZvcm1hdFRpbWUlN0QlMjBmcm9tJTIwJy4lMkZ1dGlscyclNUNuaW1wb3J0JTIwJTdCZXhhbXBsZSU3RCUyMGZyb20lMjAnLiUyRnByZWFjdCclNUNuJTVDbmZvcm1hdFRpbWUobmV3JTIwRGF0ZShtYXRjaC5hY3R1YWxUaW1lJTIwJTdDJTdDJTIwbWF0Y2gucHJlZGljdGVkVGltZSkpJTVDbiU1Q25leGFtcGxlKHVuZGVmaW5lZCklMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIydXRpbHMuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY29uc3QlMjBmb3JtYXRUaW1lJTIwJTNEJTIwKGRhdGUpJTIwJTNEJTNFJTVDbiUyMCUyMGRhdGUudG9Mb2NhbGVUaW1lU3RyaW5nKHVuZGVmaW5lZCUyQyUyMCU3QiU1Q24lMjAlMjAlMjAlMjBob3VyMTIlM0ElMjB0cnVlJTJDJTVDbiUyMCUyMCUyMCUyMGhvdXIlM0ElMjAnMi1kaWdpdCclMkMlNUNuJTIwJTIwJTIwJTIwbWludXRlJTNBJTIwJzItZGlnaXQnJTVDbiUyMCUyMCU3RCklMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIycHJlYWN0LmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZXhhbXBsZSUyMCUzRCUyMChuYW1lKSUyMCUzRCUzRSUyMCU3QiU1Q24lNUN0aWYlMjAobmFtZSUyMCUzRCUzRCUzRCUyMHVuZGVmaW5lZCklMjAlN0IlNUNuJTVDdCU1Q3Rjb25zb2xlLmxvZygnbmFtZSUyMGlzJTIwdW5kZWZpbmVkJyklNUNuJTVDdCU3RCUyMGVsc2UlMjAlN0IlNUNuJTVDdCU1Q3Rjb25zb2xlLmxvZygnaGknJTJDJTIwbmFtZSklNUNuJTVDdCU3RCU1Q24lN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

If `undefined` is used in multiple modules, it is renamed to undefined$0 with a number after the dollar sign to avoid naming collisions. In this case though, undefined should not be renamed.

This bug was introduced in `v0.61.0`, we didn't have this problem in `v0.60.7`

**Comment by qubyte** [comment]
Further reduction. This seems to be triggered by an `import`. Comment it out to get an unbroken `undefined`.

https://rollupjs.org/repl?version=0.61.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGhlbGxvJTIwZnJvbSUyMCcuJTJGaGknJTNCJTVDbnZhciUyMGElM0IlNUNuY29uc29sZS5sb2coYSUyMCUzRCUzRCUzRCUyMHVuZGVmaW5lZCklMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyaGkuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyJTJGJTJGJTIwZW1wdHklMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by lukastaegert** [comment]
REALLY sorry about this one. Fix at #2291, hope to be able to release it later today.

**Comment by calebeby** [comment]
Awesome, thanks!

**Comment by lukastaegert** [comment]
In order to track `undefined` as value in conditionals, we added a pseudo-variable to track it. Unfortunately, it was done on module level instead of the global level as we did not consider that in this situation, the deconflicting logic would kick in and start renaming instances if there is more than one module. And apparently nowhere in our tests, we had a situation where there was both an "undefined" and an import in the same file... 🙄

**Comment by guybedford** [comment]
@lukastaegert this may be a good reason to incorporate some smoke tests into the RollupJS release process to test building and running popular libraries.

**Comment by lukastaegert** [comment]
Done already 😜 cf. #2292

**Comment by lukastaegert** [comment]
Of course it would be nice to have more smoke tests in the future for "popular libraries" other than rollup 😉. The problem which might cause some work here, though, is that a good smoke test should have some meaningful tests to check basic functionality is available. For rollup, we already have a very detailed test suite. I remember @kzc mentioning that when developing uglify-es, rollup's test suite played an integral part in finding issues from them so we are not the first to do so...

**Comment by lukastaegert** [comment]
Fix released in v0.61.1

**Comment by kzc** [comment]
>  I remember @kzc mentioning that when developing uglify-es, rollup's test suite played an integral part for them in finding issues

It was indeed. Until TypeScript was adopted. ;-P.

There aren't many large ECMAScript 2015+ code bases with extensive test suites that are easily run. I also tested against Bublé and Butternut. I changed the Rollup config to not transpile the code and leave it in the original form and then minified it in-situ and ran "npm test". The source code had to be minified before Rollup unwound the imports and exports. Several bugs in Rollup, Bublé and Acorn were also exposed in the process - a virtuous circle.
